runtime/interpreter/value.go

```diff
@@ -2946,6 +2968,81 @@ func (v *ArrayValue) Reverse(
 	)
 }
 
+func (v *ArrayValue) Filter(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	procedure FunctionValue,
+) Value {
+	filteredValues := make([]Value, 0)
```

Comment:
Instead of allocating a temporary slice here and using `Iterate` to accumulate the matching elements, get an iterator for the atree array, then get next elements and test them in `NewArrayValueWithIterator` below. 

This avoids the up-front allocation, which can be potentially very large (and is currently unmetered)

Reply:
Can you please take a look now if the updated code is better?

We still need two iterations - first to get the count and the second to actually populate. We can't really meter without knowing the size but I was able to avoid the allocation. 

It comes at the cost of applying the predicate twice. I could avoid the double invocation by storing just the indexes in the first iteration but then it would mean we allocate (integer slice) before metering which I guess we don't want to do.